sotodlib/io/imprinter.py

```diff
@@ -682,9 +686,132 @@ def delete_book_files(self, book):
             self.logger.warning(f"Failed to remove {book.path}: {e}")
             self.logger.error(traceback.format_exc())
 
+    def get_book_times(self, book):
+        """
+        Given a book of co-sampled data with a timing system, this function will
+        return a gap-filled list of timestamps that should be contained in the book.
+
+        Parameters
+        ----------
+        book: Book object
+        """
+        filedb = self.get_files_for_book(book)
+        t0 = book.start.timestamp()
+        t1 = book.stop.timestamp()
+        for i, files in enumerate(filedb.values()):
+            _files = sorted(files)
+            if i == 0:  # Get full list of timestamps for first file list
+                ts = []
+                for frame in itertools.chain(*[core.G3File(f) for f in _files]):
+                    if frame.type != core.G3FrameType.Scan:
+                        continue
+                    ts.append(get_timestamps(frame))
+                ts = fill_gaps(np.hstack(ts))
+
+                _t0, _t1 = ts[[0, -1]]
+            else:
+                _t0, _t1 = get_start_and_end(_files)
+            
+            t0 = max(t0, _t0)
```

Comment:
I'm not sure we want to use book.start and book.stop to inform the min and max here, with the same reasoning as in the HWP PR that these timestamps are not accurate. You should only trust the timestamps from frames, so just point t0 and t1 to _t0 and _t1 in the first wafer would suffice.

Reply:
I kind of thought this would be a good chance to enforce this, so as it is now the data will be trimmed so that all samples are within `book.start` and `book.stop`. I'm also fine with removing this though, but then we will always need to keep in mind that the `start` and `stop` are not hard limits.

Reply:
How about replacing book.start and book.end with the correct timestamps from this function?

Reply:
I'll just leave it for now and not cut based on the book `start` and `stop`. I do think we should think about using the `get_start_and_end` function to set book.start and book.stop when the book is created though.